lib/sycamore/sycamore/transforms/detr_partitioner.py

```diff
@@ -385,108 +379,99 @@ def _partition_pdf_batched_named(
-                for stat in top_stats[:10]:
-                    print(stat)
-                before = after
-                display_top(after)
         if tracemalloc.is_tracing():
+            gc.collect()
             (current, peak) = tracemalloc.get_traced_memory()
             logger.info(f"Memory Usage current={current} peak={peak}")
-            top = tracemalloc.take_snapshot()
-            display_top(top)
+            after = tracemalloc.take_snapshot()
+            top_stats = after.compare_to(before, "lineno")
+
+            print("[ Top 10 differences ]")
+            for stat in top_stats[:10]:
+                print(stat)
+            before = after
+            display_top(after)
         return deformable_layout
 
-    @staticmethod
-    def _run_pdfminer(pdf_path, hash_key, use_cache):
-        pdfminer = PDFMinerExtractor()
-        with LogTime("pdfminer_extract", log_start=True):
-            pdfminer_layout = pdfminer.extract_document(pdf_path, hash_key, use_cache)
-
-        return pdfminer_layout
-
-    @requires_modules("easyocr", extra="local-inference")
     def process_batch(
         self,
         batch: list[Image.Image],
+        threshold: float,
+        text_extractor: TextExtractor,
         pdf_generator,
-        threshold,
         use_ocr,
         ocr_images,
-        ocr_tables,
+        ocr_model,
+        per_element_ocr,
         extract_table_structure,
         table_structure_extractor,
         extract_images,
         use_cache,
     ) -> Any:
-        import easyocr
-
-        self._init_model()
-
         with LogTime("infer"):
             assert self.model is not None
             deformable_layout = self.model.infer(batch, threshold, use_cache)
 
         gc_tensor_dump()
         assert len(deformable_layout) == len(batch)
-        if use_ocr:
-            with LogTime("ocr"):
-                if self.ocr_table_reader is None:
-                    self.ocr_table_reader = easyocr.Reader(["en"])
-
+        if use_ocr and per_element_ocr:
+            with LogTime("Per Element OCR"):
                 extract_ocr(
                     batch,
                     deformable_layout,
                     ocr_images=ocr_images,
-                    ocr_tables=ocr_tables,
-                    table_reader=self.ocr_table_reader,
+                    ocr_model_name=ocr_model,
                 )
         else:
-            pdfminer_pages = []
-            for _ in range(len(batch)):
-                pdf_page = pdf_generator.__next__()
-                pdfminer_pages.append(self.pdf_extractor.extract_page(pdf_page))
-
-            assert len(pdfminer_pages) == len(deformable_layout)
-            with LogTime("pdfminer_supplement"):
-                for d, p in zip(deformable_layout, pdfminer_pages):
+            extracted_pages = []
+            with LogTime("text_extraction"):
+                for i in range(len(batch)):
+                    page = batch[i] if use_ocr else pdf_generator.__next__()
```

Comment:
This is a weird construct.  I'd move the testing of `use_ocr` up and have two simpler loops.  It would be good if you didn't need to call `__next__` directly.  Also, I'm not sure if you have to deal with this, but what if the generator doesn't yield the same number of pages as batch?